"node:timers not found" error during build

The `node:timers` package doesn't appear to have been installed properly, causing some builds to fail. This behavior was introduced [here](https://github.com/heroku/buildpacks-nodejs/pull/742/files) in v2.6.1.

Specifically,

```
Error: Cannot find module 'node:timers'
Require stack:
- /layers/heroku_nodejs-engine/node_runtime_metrics/metrics_collector.cjs
- internal/preload
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/layers/heroku_nodejs-engine/node_runtime_metrics/metrics_collector.cjs:24:25)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/layers/heroku_nodejs-engine/node_runtime_metrics/metrics_collector.cjs',
    'internal/preload'
  ]
```

Is there a way to disable this by default? Or maybe debug why the package isn't being found?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -122,7 +122,9 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        version: [ latest, lts/*, lts/-1 ]
+        # check the minimum node version supported by the metrics script and the latest node version
+        # (assumes the versions between have backwards-compatible APIs)
+        version: [14.10.0, latest]
     name: Test Metrics (${{ matrix.version }})
     runs-on: ubuntu-latest
     steps:
@@ -130,4 +132,4 @@ jobs:
       - uses: actions/setup-node@v4
         with:
           node-version: ${{ matrix.version }}
-      - run: node --test buildpacks/nodejs-engine/node_runtime_metrics/test/metrics.spec.cjs
+      - run: npx mocha buildpacks/nodejs-engine/node_runtime_metrics/test/metrics.spec.cjs

--- a/buildpacks/nodejs-engine/CHANGELOG.md
+++ b/buildpacks/nodejs-engine/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Collect Node.js Runtime Metrics for v14.10.0 and up if the application has opted-in. ([#767](https://github.com/heroku/buildpacks-nodejs/pull/767))
+
 ## [2.6.4] - 2024-01-17
 
 - Added Node.js version 21.6.0.

--- a/buildpacks/nodejs-engine/CHANGELOG.md
+++ b/buildpacks/nodejs-engine/CHANGELOG.md
@@ -7,8 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added 
+
 - Added Node.js version 21.6.1.
-- 
+
 ### Changed
 
 - Collect Node.js Runtime Metrics for v14.10.0 and up if the application has opted-in. ([#767](https://github.com/heroku/buildpacks-nodejs/pull/767))

--- a/buildpacks/nodejs-engine/node_runtime_metrics/metrics_collector.cjs
+++ b/buildpacks/nodejs-engine/node_runtime_metrics/metrics_collector.cjs
@@ -21,16 +21,18 @@
  *     gauges: EventLoopGauges;
  * }} MetricsPayload
  */
-const { setInterval } = require('node:timers')
-const { URL } = require('node:url');
-const { debuglog } = require('node:util');
-const { monitorEventLoopDelay, PerformanceObserver, constants, performance } = require('node:perf_hooks')
+const { setInterval } = require('timers')
+const { URL } = require('url');
+const { debuglog } = require('util');
+const { monitorEventLoopDelay, PerformanceObserver, constants, performance } = require('perf_hooks')
+const { request: insecureRequest } = require('http');
+const { request: secureRequest } = require('https');
 
 try {
     // failures from the instrumentation shouldn't mess with the application
     registerInstrumentation()
 } catch (e) {
-    log(`An unexpected error occurred: ${e.message}`)
+    log(`An unexpected error occurred: ${e.stack}`)
 }
 
 /**
@@ -53,7 +55,7 @@ function registerInstrumentation() {
     const gcObserver = new PerformanceObserver((value) => {
         value.getEntries().forEach(entry => updateMemoryCounters(memoryCounters, entry))
     })
-    gcObserver.observe({ type: 'gc' })
+    gcObserver.observe({ entryTypes: ['gc'] })
 
     const eventLoopHistogram = monitorEventLoopDelay()
     eventLoopHistogram.enable()
@@ -61,25 +63,25 @@ function registerInstrumentation() {
     let previousEventLoopUtilization = performance.eventLoopUtilization()
 
     const timeout  = setInterval(() => {
-        const eventLoopUtilization = performance.eventLoopUtilization(previousEventLoopUtilization)
-        eventLoopHistogram.disable()
-        gcObserver.disconnect()
+        try {
+            const eventLoopUtilization = performance.eventLoopUtilization(previousEventLoopUtilization)
+            eventLoopHistogram.disable()
+            gcObserver.disconnect()
 
-        const payload = {
-            counters: { ...memoryCounters },
-            gauges: captureEventLoopGauges(eventLoopUtilization, eventLoopHistogram)
+            sendMetrics(herokuMetricsUrl, {
+                counters: {...memoryCounters},
+                gauges: captureEventLoopGauges(eventLoopUtilization, eventLoopHistogram)
+            })
+
+            // reset memory and event loop measures
+            previousEventLoopUtilization = eventLoopUtilization
+            memoryCounters = initializeMemoryCounters()
+            gcObserver.observe({ entryTypes: ['gc'] })
+            eventLoopHistogram.reset()
+            eventLoopHistogram.enable()
+        } catch (e) {
+            log(`An unexpected error occurred: ${e.stack}`)
         }
-
-        sendMetrics(herokuMetricsUrl, payload).catch((err) => {
-            log(`An error occurred while sending metrics - ${err}`)
-        })
-
-        // reset memory and event loop measures
-        previousEventLoopUtilization = eventLoopUtilization
-        memoryCounters = initializeMemoryCounters()
-        gcObserver.observe({ type: 'gc' })
-        eventLoopHistogram.reset()
-        eventLoopHistogram.enable()
     }, herokuMetricsInterval)
 
     // `setInterval` actually returns a Timeout object but this isn't recognized by the type-checker which
@@ -171,19 +173,31 @@ function initializeMemoryCounters(){
  * @param {PerformanceEntry} performanceEntry
  */
 function updateMemoryCounters(memoryCounters, performanceEntry) {
-    // only record entries with that contain a 'detail' object with 'kind' property
-    // since these are only available for NodeGCPerformanceDetail entries
-    if ('detail' in performanceEntry && 'kind' in performanceEntry.detail) {
-        const nsDuration = millisecondsToNanoseconds(performanceEntry.duration)
-        memoryCounters['node.gc.collections'] += 1
-        memoryCounters['node.gc.pause.ns'] += nsDuration
-        if (performanceEntry.detail.kind === constants.NODE_PERFORMANCE_GC_MINOR) {
-            memoryCounters['node.gc.young.collections'] += 1
-            memoryCounters['node.gc.young.pause.ns'] += nsDuration
-        } else {
-            memoryCounters['node.gc.old.collections'] += 1
-            memoryCounters['node.gc.old.pause.ns'] += nsDuration
-        }
+    const nsDuration = millisecondsToNanoseconds(performanceEntry.duration)
+    memoryCounters['node.gc.collections'] += 1
+    memoryCounters['node.gc.pause.ns'] += nsDuration
+    if (getGcPerformanceEntryKind(performanceEntry) === constants.NODE_PERFORMANCE_GC_MINOR) {
+        memoryCounters['node.gc.young.collections'] += 1
+        memoryCounters['node.gc.young.pause.ns'] += nsDuration
+    } else {
+        memoryCounters['node.gc.old.collections'] += 1
+        memoryCounters['node.gc.old.pause.ns'] += nsDuration
+    }
+}
+
+/**
+ * Reads the `kind` field for the 'gc' performance entry in a backwards-compatible way
+ * @param {PerformanceEntry} performanceEntry
+ * @returns {number}
+ */
+function getGcPerformanceEntryKind(performanceEntry) {
+    // using try/catch to avoid triggering deprecation warnings
+    try {
+        // for v16 and up
+        return performanceEntry.detail.kind
+    } catch (e) {
+        // fallback for v14 & v15
+        return performanceEntry.kind
     }
 }
 
@@ -209,7 +223,7 @@ function captureEventLoopGauges(eventLoopUtilization, eventLoopHistogram) {
  * @return {number}
  */
 function millisecondsToNanoseconds(ms) {
-    return ms * 1_000_000
+    return ms * 1e6 // 1_000_000
 }
 
 /**
@@ -218,26 +232,47 @@ function millisecondsToNanoseconds(ms) {
  * @return {number}
  */
 function nanosecondsToMilliseconds(ns) {
-    return ns / 1_000_000
+    return ns / 1e6 // 1_000_000
 }
 
 /**
  * Sends the collected metrics to the given endpoint using a POST request.
  * @param {URL} url
  * @param {MetricsPayload} payload
- * @returns {Promise<void>}
+ * @returns void
  */
 function sendMetrics(url, payload) {
+    const request = url.protocol === 'https:' ? secureRequest : insecureRequest
+    const payloadAsJson = JSON.stringify(payload)
+
     log(`Sending metrics to ${url.toString()}`)
-    return fetch(url, {
+    const clientRequest = request({
         method: 'POST',
-        headers: { 'Content-Type': 'application/json' },
-        body: JSON.stringify(payload)
-    }).then(res => {
-        if (res.ok) {
-            log('Metrics sent successfully')
-        } else {
-            log(`Tried to send metrics but response was: ${res.status} - ${res.statusText}`)
+        protocol: url.protocol,
+        hostname: url.hostname,
+        port: url.port,
+        path: url.pathname,
+        headers: {
+            'Content-Type': 'application/json',
+            'Content-Length': Buffer.byteLength(payloadAsJson)
         }
     })
+
+    clientRequest.on('response', (res) => {
+        if (res.statusCode === 200) {
+            log('Metrics sent successfully')
+        } else {
+            log(`Tried to send metrics but response was: ${res.statusCode} - ${res.statusMessage}`)
+        }
+        // consume response data to free up memory
+        // see: https://nodejs.org/docs/latest/api/http.html#http_class_http_clientrequest
+        res.resume()
+    })
+
+    clientRequest.on('error', (err) => {
+        log(`An error occurred while sending metrics - ${err}`)
+    })
+
+    clientRequest.write(payloadAsJson)
+    clientRequest.end()
 }

--- a/buildpacks/nodejs-engine/node_runtime_metrics/test/fixtures/clustered_app.cjs
+++ b/buildpacks/nodejs-engine/node_runtime_metrics/test/fixtures/clustered_app.cjs
@@ -1,10 +1,13 @@
-const { isPrimary, fork } = require('node:cluster')
+const cluster = require('cluster')
 
 require('./_cpu_and_memory_simulator.cjs')
 
-if (isPrimary) {
+if (
+    cluster.isPrimary ||
+    cluster.isMaster // deprecated in Node v16
+) {
     console.log(`Starting primary cluster ${process.pid} running with NODE_OPTIONS="${process.env.NODE_OPTIONS || ''}"`)
-    fork()
+    cluster.fork()
     process.on('SIGINT', () => {
         process.exit(0)
     })

--- a/buildpacks/nodejs-engine/node_runtime_metrics/test/fixtures/worker_threads_app.cjs
+++ b/buildpacks/nodejs-engine/node_runtime_metrics/test/fixtures/worker_threads_app.cjs
@@ -1,4 +1,4 @@
-const { Worker, isMainThread} = require('node:worker_threads')
+const { Worker, isMainThread} = require('worker_threads')
 
 require('./_cpu_and_memory_simulator.cjs')
 

--- a/buildpacks/nodejs-engine/node_runtime_metrics/test/metrics.spec.cjs
+++ b/buildpacks/nodejs-engine/node_runtime_metrics/test/metrics.spec.cjs
@@ -1,29 +1,31 @@
-const { existsSync} = require('node:fs')
-const { join, resolve, dirname } = require('node:path')
-const { spawn } = require('node:child_process')
-const http = require('node:http')
-const { afterEach, beforeEach, describe, it } = require('node:test')
-const assert = require("node:assert");
+/**
+ * This test suite can be executed with `npx mocha ./metrics.spec.cjs`. Initially it was meant to run with the built-in
+ * Node.js test runner but now that support is being expanded to Node.js v14.10.0 and up we need an external test runner
+ * and so `mocha` + the built-in assertion library in Node.js is all we need.
+ */
+
+const { existsSync} = require('fs')
+const { join, resolve, dirname } = require('path')
+const { spawn } = require('child_process')
+const http = require('http')
+const assert = require('assert');
 
 const metricsScript = resolve(__dirname, '..', 'metrics_collector.cjs')
 
-const testOptions = {
-    // the default metric interval is 20 seconds, so we'll make these tests run just slightly longer than
-    // the time required to collect a single metric
-    timeout: 21000,
-}
+// the default metric interval is 20 seconds, so we'll make these tests run just slightly longer than
+// the time required to collect a single metric
+const TEST_TIMEOUT = 21000
 
-describe('Metrics plugin', () => {
-    let abortController
+describe('Metrics plugin', function () {
+    // because we use `this` here the above describe needs to use `function () { ... }` instead of `() => { ... }`
+    this.timeout(TEST_TIMEOUT)
 
-    beforeEach(() => {
-        // this is used to shut down the internal web server that receives metrics
-        abortController = new AbortController()
-    })
+    let metricsReceiver
+    let application
 
     afterEach( () => {
-        if (abortController) {
-            abortController.abort('Test Complete')
+        if (metricsReceiver) {
+            metricsReceiver.disconnect()
         }
     })
 
@@ -59,9 +61,9 @@ describe('Metrics plugin', () => {
             }
         }
 
-        it('should collect metrics from an application running in a single process', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should collect metrics from an application running in a single process', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 10000 // let's collect more than one metric
             })
@@ -69,9 +71,9 @@ describe('Metrics plugin', () => {
             assertMetricsReceived(metricsReceiver.metricsReceived, 2)
         })
 
-        it('should support worker threads', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('worker_threads_app.cjs', {
+        it('should support worker threads', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('worker_threads_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 10000 // let's collect more than one metric
             })
@@ -79,9 +81,9 @@ describe('Metrics plugin', () => {
             assertMetricsReceived(metricsReceiver.metricsReceived, 4)
         })
 
-        it('should support clustering', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('clustered_app.cjs', {
+        it('should support clustering', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('clustered_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 10000 // let's collect more than one metric
             })
@@ -91,8 +93,8 @@ describe('Metrics plugin', () => {
     })
 
     describe('configuration', () => {
-        it('should exit if no metrics url is provided', testOptions, async () => {
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should exit if no metrics url is provided', async () => {
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: undefined,
                 msToExecute: 1000
             })
@@ -100,8 +102,8 @@ describe('Metrics plugin', () => {
             assert.match(application.pluginOutput, /\[heroku-metrics] Metrics will not be collected for this application/)
         })
 
-        it('should exit if metrics url is invalid', testOptions, async () => {
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should exit if metrics url is invalid', async () => {
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: 'not a url',
                 msToExecute: 1000
             })
@@ -109,9 +111,9 @@ describe('Metrics plugin', () => {
             assert.match(application.pluginOutput, /\[heroku-metrics] Invalid URL:/)
         })
 
-        it('should use a default interval of 20 seconds', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should use a default interval of 20 seconds', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: undefined,
                 msToExecute: 1000
@@ -119,9 +121,9 @@ describe('Metrics plugin', () => {
             assert.match(application.pluginOutput, /\[heroku-metrics] Using default interval of 20000ms/)
         })
 
-        it('should allow the default interval to be changed', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should allow the default interval to be changed', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 10000,
                 msToExecute: 1000
@@ -130,9 +132,9 @@ describe('Metrics plugin', () => {
             assert.match(application.pluginOutput, /\[heroku-metrics] Using interval of 10000ms/)
         })
 
-        it('should not allow the default interval to be changed to less than 10 seconds', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should not allow the default interval to be changed to less than 10 seconds', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 9999,
                 msToExecute: 1000
@@ -141,9 +143,9 @@ describe('Metrics plugin', () => {
             assert.match(application.pluginOutput, /\[heroku-metrics] Interval is lower than the minimum, using the minimum interval of 10000ms instead/)
         })
 
-        it('should not allow the default interval to be changed to by a non-numeric value', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController)
-            const application = await spawnApplication('single_process_app.cjs', {
+        it('should not allow the default interval to be changed to by a non-numeric value', async () => {
+            metricsReceiver = await startMetricsReceiver()
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url,
                 metricsIntervalOverride: 'not a number',
                 msToExecute: 1000
@@ -154,11 +156,11 @@ describe('Metrics plugin', () => {
     })
 
     describe('http error', () => {
-        it('should report when requests fail', testOptions, async () => {
-            const metricsReceiver = await startMetricsReceiver(abortController, {
+        it('should report when requests fail', async () => {
+            metricsReceiver = await startMetricsReceiver( {
                 responseStatusCode: 429
             })
-            const application = await spawnApplication('single_process_app.cjs', {
+            application = await spawnApplication('single_process_app.cjs', {
                 metricsUrl: metricsReceiver.url
             })
             assert.match(application.pluginOutput, /\[heroku-metrics] Tried to send metrics but response was: 429 - Too Many Requests/)
@@ -166,7 +168,7 @@ describe('Metrics plugin', () => {
     })
 })
 
-function startMetricsReceiver(abortController, options = {}) {
+function startMetricsReceiver(options = {}) {
     const metricsReceived = []
     options = {
         responseStatusCode: 200,
@@ -186,13 +188,13 @@ function startMetricsReceiver(abortController, options = {}) {
 
         metricsReceiver.listen({
             port: 0, // auto-assign the port
-            signal: abortController.signal
         }, () => {
             const port = metricsReceiver.address().port
             console.log(`- metrics receiver listening on ${port}`)
             resolve({
                 url: `http://localhost:${port}`,
-                metricsReceived
+                metricsReceived,
+                disconnect: () => metricsReceiver.close()
             })
         })
 
@@ -214,7 +216,7 @@ function spawnApplication(fixture, options = {}) {
     options = {
         metricsUrl: undefined,
         metricsIntervalOverride: undefined,
-        msToExecute: testOptions.timeout - 200, // kill the service just before the timeout is reached
+        msToExecute: TEST_TIMEOUT - 200, // kill the service just before the timeout is reached (so tests don't hang)
         ...options
     }
 
@@ -249,9 +251,15 @@ function spawnApplication(fixture, options = {}) {
             process.stdout.write(d)
         })
 
-        metricsProducer.on('close', (code) => {
+        metricsProducer.once('close', (code) => {
+            if (!code) {
+                code = 0
+            }
             if (code === 0 || code === 130) {
-                resolve({code, pluginOutput})
+                resolve({
+                    code,
+                    pluginOutput
+                })
             } else {
                 console.log(`EXIT: ${code}`)
                 console.log(`STDERR:\n${pluginOutput}`)
@@ -259,7 +267,7 @@ function spawnApplication(fixture, options = {}) {
             }
         })
 
-        // kill the service just before the timeout is reached
+        // run the application for the given time and then kill it
         setTimeout(() => {
             metricsProducer.kill('SIGINT') // code 130
         }, options.msToExecute)

--- a/buildpacks/nodejs-engine/src/main.rs
+++ b/buildpacks/nodejs-engine/src/main.rs
@@ -29,6 +29,8 @@ const INVENTORY: &str = include_str!("../inventory.toml");
 
 const LTS_VERSION: &str = "20.x";
 
+const MINIMUM_NODE_VERSION_FOR_METRICS: &str = ">=14.10";
+
 struct NodeJsEngineBuildpack;
 
 impl Buildpack for NodeJsEngineBuildpack {
@@ -103,7 +105,13 @@ impl Buildpack for NodeJsEngineBuildpack {
         )?;
 
         context.handle_layer(layer_name!("web_env"), WebEnvLayer)?;
-        context.handle_layer(layer_name!("node_runtime_metrics"), NodeRuntimeMetricsLayer)?;
+
+        if Requirement::parse(MINIMUM_NODE_VERSION_FOR_METRICS)
+            .expect("should be a valid version range")
+            .satisfies(&target_release.version)
+        {
+            context.handle_layer(layer_name!("node_runtime_metrics"), NodeRuntimeMetricsLayer)?;
+        }
 
         let launchjs = ["server.js", "index.js"]
             .map(|name| context.app_dir.join(name))

--- a/buildpacks/nodejs-engine/tests/integration_test.rs
+++ b/buildpacks/nodejs-engine/tests/integration_test.rs
@@ -1,7 +1,7 @@
 // Required due to: https://github.com/rust-lang/rust/issues/95513
 #![allow(unused_crate_dependencies)]
 
-use libcnb_test::{assert_contains, ContainerConfig};
+use libcnb_test::{assert_contains, assert_not_contains, ContainerConfig};
 use std::time::Duration;
 use test_support::{
     assert_web_response, nodejs_integration_test, nodejs_integration_test_with_config,
@@ -100,4 +100,53 @@ fn runtime_metrics_script_is_not_activated_when_heroku_metrics_url_is_not_set() 
             );
         });
     });
+}
+
+#[test]
+#[ignore]
+fn runtime_metrics_script_is_activated_when_node_version_is_at_least_v14_10_0() {
+    nodejs_integration_test_with_config(
+        "./fixtures/node-with-indexjs",
+        |config| {
+            config.app_dir_preprocessor(|app_dir| {
+                set_node_engine(&app_dir, "14.10.0");
+            });
+        },
+        |ctx| {
+            let mut container_config = ContainerConfig::new();
+            container_config
+                .expose_port(PORT)
+                .env("NODE_DEBUG", "heroku")
+                .env("HEROKU_METRICS_URL", "http://localhost:3000");
+
+            ctx.start_container(container_config, |container| {
+                let output = container.logs_now();
+                assert_contains!(output.stderr, "Registering metrics instrumentation");
+            });
+        },
+    );
+}
+
+#[test]
+#[ignore]
+fn runtime_metrics_script_is_not_activated_when_node_version_is_less_than_v14_10_0() {
+    nodejs_integration_test_with_config(
+        "./fixtures/node-with-indexjs",
+        |config| {
+            config.app_dir_preprocessor(|app_dir| {
+                set_node_engine(&app_dir, "14.9.0");
+            });
+        },
+        |ctx| {
+            let mut container_config = ContainerConfig::new();
+            container_config
+                .expose_port(PORT)
+                .env("NODE_DEBUG", "heroku");
+
+            ctx.start_container(container_config, |container| {
+                let output = container.logs_now();
+                assert_not_contains!(output.stderr, "Registering metrics instrumentation");
+            });
+        },
+    );
 }

--- a/buildpacks/nodejs-engine/tests/integration_test.rs
+++ b/buildpacks/nodejs-engine/tests/integration_test.rs
@@ -120,6 +120,7 @@ fn runtime_metrics_script_is_activated_when_node_version_is_at_least_v14_10_0() 
                 .env("HEROKU_METRICS_URL", "http://localhost:3000");
 
             ctx.start_container(container_config, |container| {
+                std::thread::sleep(Duration::from_secs(1));
                 let output = container.logs_now();
                 assert_contains!(output.stderr, "Registering metrics instrumentation");
             });

--- a/common/nodejs-utils/src/vrs.rs
+++ b/common/nodejs-utils/src/vrs.rs
@@ -104,7 +104,7 @@ impl Requirement {
     }
 
     #[must_use]
-    pub(crate) fn satisfies(&self, ver: &Version) -> bool {
+    pub fn satisfies(&self, ver: &Version) -> bool {
         self.0.satisfies(&ver.0)
     }
 }

